Restructure final loop of distributeCores to iterate over threadsToReceiveCores

hq6 opened the pull request.

Commit:
Restructure final loop of distributeCorest

The loop that actually assigns cores to threads now iterates over
threadsToReceiveCores instead of managedCores.

This is the first step in allowing threads to prefer certain cores over others,
for the purposes of keeping threads of the same application on hypertwins of
the same physical core.

The original code iterates over cores and puts a thread on them, picking the
threads in priority order.  If we continued to iterate over cores and pick
threads to run on them, we may violate priority order.

To preserve priority order while introducing affinity for cores, we must
instead iterate over threads and find a core to put them on. Later commits can
change the mechanism for finding a core without violating priority correctness.

jspeiser reviewed (approved):
I think it looks good!

Commit:
Add optionally compiled pauses to CoreArbiterRequestTest

Commit:
Add loose affinity for hypertwins

Managed cores are still considered before unmanaged cores, and core
deallocation order is still arbitrary and out of the arbiter's control.

Commit:
Revert "Update CoreArbiterServerMain to accept the allAvailableCores option"

This reverts commit 0fe830bd221dead3f44d2dce027d3acdd95db05e.

Commit:
Revert "Add the concept of all cores to the CoreArbiterServer"

This reverts commit 099e0107ebd7a85c2933e4fe7d264359faf4cf07.

Commit:
Add a test for cores not ramping down

Commit:
Prempt all pre-emptible cores when distributing cores.

This ensures that requests to reduce the number of cores for a given
application will be honored even if there are no other Arachne applications
requesting cores.

hq6 commented:
This PR has been subsumed by a subsequent PR.
